File: src/ee/facebook_ads/private/FacebookAdsBridge.swift
//
//  FacebookAds.swift
//  Pods
//
//  Created by eps on 6/25/20.
//

import FBAudienceNetwork
import Foundation

private let kPrefix = "FacebookAdsBridge"
private let kGetTestDeviceHash = "\(kPrefix)GetTestDeviceHash"
private let kAddTestDevice = "\(kPrefix)AddTestDevice"
private let kClearTestDevices = "\(kPrefix)ClearTestDevices"
private let kGetBannerAdSize = "\(kPrefix)GetBannerAdSize"
private let kCreateBannerAd = "\(kPrefix)CreateBannerAd"
private let kDestroyBannerAd = "\(kPrefix)DestroyBannerAd"
private let kCreateNativeAd = "\(kPrefix)CreateNativeAd"
private let kDestroyNativeAd = "\(kPrefix)DestroyNativeAd"
private let kCreateInterstitialAd = "\(kPrefix)CreateInterstitialAd"
private let kDestroyInterstitialAd = "\(kPrefix)DestroyInterstitialAd"
private let kCreateRewardedAd = "\(kPrefix)CreateRewardedAd"
private let kDestroyRewardedAd = "\(kPrefix)DestroyRewardedAd"

@objc(EEFacebookAdsBridge)
public class FacebookAdsBridge: NSObject, IPlugin {
    private let _bridge: IMessageBridge
    private let _bannerHelper = FacebookBannerHelper()
    private var _bannerAds: [String: FacebookBannerAd] = [:]
    private var _nativeAds: [String: FacebookNativeAd] = [:]
    private var _interstitialAds: [String: FacebookInterstitialAd] = [:]
    private var _rewardedAds: [String: FacebookRewardedAd] = [:]

    public required init(_ bridge: IMessageBridge) {
        _bridge = bridge
        super.init()
        registerHandlers()
    }

    public func destroy() {
        deregisterHandlers()
        _bannerAds.values.forEach { $0.destroy() }
        _bannerAds.removeAll()
        _nativeAds.values.forEach { $0.destroy() }
        _nativeAds.removeAll()
        _interstitialAds.values.forEach { $0.destroy() }
        _interstitialAds.removeAll()
        _rewardedAds.values.forEach { $0.destroy() }
        _rewardedAds.removeAll()
    }

    func registerHandlers() {
        _bridge.registerHandler(kGetTestDeviceHash) { _ in
            self.testDeviceHash
        }
        _bridge.registerHandler(kAddTestDevice) { message in
            self.addTestDevice(message)
            return ""
        }
        _bridge.registerHandler(kClearTestDevices) { _ in
            self.clearTestDevice()
            return ""
        }
        _bridge.registerHandler(kGetBannerAdSize) { message in
            let sizeId = Int(message) ?? -1
            let size = self.getBannerAdSize(sizeId)
            return EEJsonUtils.convertDictionary(toString: [
                "width": size.width,
                "height": size.height
            ])
        }
        _bridge.registerHandler(kCreateBannerAd) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard let adId = dict["ad_id"] as? String else {
                assert(false, "Invalid argument")
            }
            guard let sizeId = dict["ad_size"] as? Int else {
                assert(false, "Invalid argument")
            }
            let adSize = self._bannerHelper.getAdSize(sizeId)
            return Utils.toString(self.createBannerAd(adId, adSize))
        }
        _bridge.registerHandler(kDestroyBannerAd) { message in
            Utils.toString(self.destroyBannerAd(message))
        }
        _bridge.registerHandler(kCreateNativeAd) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard let adId = dict["ad_id"] as? String else {
                assert(false, "Invalid argument")
            }
            guard let layoutName = dict["layout_name"] as? String else {
                assert(false, "Invalid argument")
            }
            return Utils.toString(self.createNativeAd(adId, layoutName))
        }
        _bridge.registerHandler(kDestroyNativeAd) { message in
            Utils.toString(self.destroyNativeAd(message))
        }
        _bridge.registerHandler(kCreateInterstitialAd) { message in
            Utils.toString(self.createInterstitialAd(message))
        }
        _bridge.registerHandler(kDestroyInterstitialAd) { message in
            Utils.toString(self.destroyInterstitialAd(message))
        }
        _bridge.registerHandler(kCreateRewardedAd) { message in
            Utils.toString(self.createRewardedAd(message))
        }
        _bridge.registerHandler(kDestroyRewardedAd) { message in
            Utils.toString(self.destroyRewardedAd(message))
        }
    }

    func deregisterHandlers() {
        _bridge.deregisterHandler(kGetTestDeviceHash)
        _bridge.deregisterHandler(kAddTestDevice)
        _bridge.deregisterHandler(kClearTestDevices)
        _bridge.deregisterHandler(kGetBannerAdSize)
        _bridge.deregisterHandler(kCreateBannerAd)
        _bridge.deregisterHandler(kDestroyBannerAd)
        _bridge.deregisterHandler(kCreateNativeAd)
        _bridge.deregisterHandler(kDestroyNativeAd)
        _bridge.deregisterHandler(kCreateInterstitialAd)
        _bridge.deregisterHandler(kDestroyInterstitialAd)
        _bridge.deregisterHandler(kCreateRewardedAd)
        _bridge.deregisterHandler(kDestroyRewardedAd)
    }

    var testDeviceHash: String {
        return FBAdSettings.testDeviceHash()
    }

    func addTestDevice(_ hash: String) {
        Thread.runOnMainThread {
            FBAdSettings.addTestDevice(hash)
        }
    }

    func clearTestDevice() {
        Thread.runOnMainThread {
            FBAdSettings.clearTestDevices()
        }
    }

    func getBannerAdSize(_ sizeId: Int) -> CGSize {
        return _bannerHelper.getSize(sizeId)
    }

    func createBannerAd(_ adId: String, _ size: FBAdSize) -> Bool {
        if _bannerAds.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = FacebookBannerAd(_bridge, adId, size)
        _bannerAds[adId] = ad
        return true
    }

    func destroyBannerAd(_ adId: String) -> Bool {
        guard let ad = _bannerAds[adId] else {
            return false
        }
        ad.destroy()
        _bannerAds.removeValue(forKey: adId)
        return true
    }

    func createNativeAd(_ adId: String, _ layoutName: String) -> Bool {
        if _nativeAds.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = FacebookNativeAd(_bridge, adId, layoutName)
        _nativeAds[adId] = ad
        return true
    }

    func destroyNativeAd(_ adId: String) -> Bool {
        guard let ad = _nativeAds[adId] else {
            return false
        }
        ad.destroy()
        _nativeAds.removeValue(forKey: adId)
        return true
    }

    func createInterstitialAd(_ adId: String) -> Bool {
        if _interstitialAds.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = FacebookInterstitialAd(_bridge, adId)
        _interstitialAds[adId] = ad
        return true
    }

    func destroyInterstitialAd(_ adId: String) -> Bool {
        guard let ad = _interstitialAds[adId] else {
            return false
        }
        ad.destroy()
        _interstitialAds.removeValue(forKey: adId)
        return true
    }

    func createRewardedAd(_ adId: String) -> Bool {
        if _rewardedAds.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = FacebookRewardedAd(_bridge, adId)
        _rewardedAds[adId] = ad
        return true
    }

    func destroyRewardedAd(_ adId: String) -> Bool {
        guard let ad = _rewardedAds[adId] else {
            return false
        }
        ad.destroy()
        _rewardedAds.removeValue(forKey: adId)
        return true
    }
}

File: src/cpp/ee/facebook_ads/private/FacebookAdsBridge.cpp
//
//  FacebookAds.cpp
//  ee_x
//
//  Created by Pham Xuan Han on 5/12/17.
//
//

#include "ee/facebook_ads/private/FacebookAdsBridge.hpp"

#include <ee/ads/internal/DefaultBannerAd.hpp>
#include <ee/ads/internal/DefaultFullScreenAd.hpp>
#include <ee/ads/internal/GuardedBannerAd.hpp>
#include <ee/ads/internal/GuardedFullScreenAd.hpp>
#include <ee/ads/internal/MediationManager.hpp>
#include <ee/core/ILogger.hpp>
#include <ee/core/IMessageBridge.hpp>
#include <ee/core/Task.hpp>
#include <ee/core/Utils.hpp>
#include <ee/nlohmann/json.hpp>

#include "ee/facebook_ads/FacebookNativeAdLayout.hpp"

namespace ee {
namespace facebook_ads {
namespace {
// clang-format off
const std::string kPrefix         = "FacebookAdsBridge";

const auto kInitialize            = kPrefix + "Initialize";

const auto kGetTestDeviceHash     = kPrefix + "GetTestDeviceHash";
const auto kAddTestDevice         = kPrefix + "AddTestDevice";
const auto kClearTestDevices      = kPrefix + "ClearTestDevices";

const auto kGetBannerAdSize       = kPrefix + "GetBannerAdSize";
const auto kCreateBannerAd        = kPrefix + "CreateBannerAd";
const auto kDestroyBannerAd       = kPrefix + "DestroyBannerAd";

const auto kCreateNativeAd        = kPrefix + "CreateNativeAd";
const auto kDestroyNativeAd       = kPrefix + "DestroyNativeAd";

const auto kCreateInterstitialAd  = kPrefix + "CreateInterstitialAd";
const auto kDestroyInterstitialAd = kPrefix + "DestroyInterstitialAd";
    
const auto kCreateRewardedAd      = kPrefix + "CreateRewardedAd";
const auto kDestroyRewardedAd     = kPrefix + "DestroyRewardedAd";
} // namespace

namespace {
// clang-format off
constexpr auto k__ad_id       = "adId";
constexpr auto k__ad_size     = "adSize";
constexpr auto k__layout_name = "layoutName";
constexpr auto k__identifiers = "identifiers";
// clang-format on
} // namespace

using Self = Bridge;

Self::Bridge(IMessageBridge& bridge, ILogger& logger,
             const Destroyer& destroyer)
    : bridge_(bridge)
    , logger_(logger)
    , destroyer_(destroyer) {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    auto&& mediation = ads::MediationManager::getInstance();
    displayer_ = mediation.getAdDisplayer();
}

Self::~Bridge() = default;

void Self::destroy() {
    logger_.debug("%s", __PRETTY_FUNCTION__);
    for (auto&& [key, value] : ads_) {
        value->destroy();
    }
    destroyer_();
}

Task<bool> Self::initialize() {
    auto response = co_await bridge_.callAsync(kInitialize);
    co_return core::toBool(response);
}

std::string Self::getTestDeviceHash() const {
    return bridge_.call(kGetTestDeviceHash);
}

void Self::addTestDevice(const std::string& hash) {
    bridge_.call(kAddTestDevice, hash);
}

void Self::clearTestDevices() {
    bridge_.call(kClearTestDevices);
}

std::pair<int, int> Self::getBannerAdSize(BannerAdSize adSize) {
    auto response = bridge_.call(kGetBannerAdSize,
                                 std::to_string(static_cast<int>(adSize)));
    auto json = nlohmann::json::parse(response);
    int width = json["width"];
    int height = json["height"];
    return std::pair(width, height);
}

std::shared_ptr<IBannerAd> Self::createBannerAd(const std::string& adId,
                                                BannerAdSize adSize) {
    logger_.debug("%s: id = %s size = %d", __PRETTY_FUNCTION__, adId.c_str(),
                  static_cast<int>(adSize));
    auto iter = ads_.find(adId);
    if (iter != ads_.cend()) {
        return std::dynamic_pointer_cast<IBannerAd>(iter->second);
    }
    nlohmann::json json;
    json[k__ad_id] = adId;
    json[k__ad_size] = static_cast<int>(adSize);
    auto response = bridge_.call(kCreateBannerAd, json.dump());
    if (not core::toBool(response)) {
        logger_.error("%s: There was an error when attempt to create an ad",
                      __PRETTY_FUNCTION__);
        assert(false);
        return nullptr;
    }
    auto size = getBannerAdSize(adSize);
    auto ad = std::make_shared<ads::GuardedBannerAd>(
        std::make_shared<ads::DefaultBannerAd>(
            "FacebookBannerAd", bridge_, logger_,
            [this, adId] { //
                destroyAd(kDestroyBannerAd, adId);
            },
            adId, size));
    ads_.emplace(adId, ad);
    return ad;
}

std::shared_ptr<IBannerAd>
Self::createNativeAd(const std::string& adId, const std::string& layoutName,
                     const NativeAdLayout& identifiers) {
    logger_.debug("%s: id = %s", __PRETTY_FUNCTION__, adId.c_str());
    auto iter = ads_.find(adId);
    if (iter != ads_.cend()) {
        return std::dynamic_pointer_cast<IBannerAd>(iter->second);
    }
    nlohmann::json json;
    json[k__ad_id] = adId;
    json[k__layout_name] = layoutName;
    json[k__identifiers] = identifiers.params_;
    auto response = bridge_.call(kCreateNativeAd, json.dump());
    if (not core::toBool(response)) {
        logger_.error("%s: There was an error when attempt to create an ad",
                      __PRETTY_FUNCTION__);
        assert(false);
        return nullptr;
    }
    auto ad = std::make_shared<ads::GuardedBannerAd>(
        std::make_shared<ads::DefaultBannerAd>(
            "FacebookNativeAd", bridge_, logger_,
            [this, adId] { //
                destroyAd(kDestroyNativeAd, adId);
            },
            adId, std::pair(0, 0)));
    ads_.emplace(adId, ad);
    return ad;
}

std::shared_ptr<IFullScreenAd>
Self::createInterstitialAd(const std::string& adId) {
    return createFullScreenAd(kCreateInterstitialAd, adId, [this, adId] {
        return std::make_shared<ads::DefaultFullScreenAd>(
            "FacebookInterstitialAd", bridge_, logger_, displayer_,
            [this, adId] { //
                return destroyAd(kDestroyInterstitialAd, adId);
            },
            [](const std::string& message) { //
                return FullScreenAdResult::Completed;
            },
            adId);
    });
}

std::shared_ptr<IFullScreenAd> Self::createRewardedAd(const std::string& adId) {
    return createFullScreenAd(kCreateRewardedAd, adId, [this, adId] {
        return std::make_shared<ads::DefaultFullScreenAd>(
            "FacebookRewardedAd", bridge_, logger_, displayer_,
            [this, adId] { //
                return destroyAd(kDestroyRewardedAd, adId);
            },
            [](const std::string& message) { //
                return core::toBool(message) ? FullScreenAdResult::Completed
                                             : FullScreenAdResult::Canceled;
            },
            adId);
    });
}

std::shared_ptr<IFullScreenAd> Self::createFullScreenAd(
    const std::string& handlerId, const std::string& adId,
    const std::function<std::shared_ptr<IFullScreenAd>()>& creator) {
    logger_.debug("%s: id = %s", __PRETTY_FUNCTION__, adId.c_str());
    auto iter = ads_.find(adId);
    if (iter != ads_.cend()) {
        return std::dynamic_pointer_cast<IFullScreenAd>(iter->second);
    }
    auto response = bridge_.call(handlerId, adId);
    if (not core::toBool(response)) {
        logger_.error("%s: There was an error when attempt to create an ad.",
                      __PRETTY_FUNCTION__);
        assert(false);
        return nullptr;
    }
    auto ad = std::make_shared<ads::GuardedFullScreenAd>(creator());
    ads_.emplace(adId, ad);
    return ad;
}

bool Self::destroyAd(const std::string& handlerId, const std::string& adId) {
    logger_.debug("%s: id = %s", __PRETTY_FUNCTION__, adId.c_str());
    auto iter = ads_.find(adId);
    if (iter == ads_.cend()) {
        return false;
    }
    auto&& response = bridge_.call(handlerId, adId);
    if (not core::toBool(response)) {
        logger_.error("%s: There was an error when attempt to destroy an ad.",
                      __PRETTY_FUNCTION__);
        assert(false);
        return false;
    }
    ads_.erase(iter);
    return true;
}
} // namespace facebook_ads
} // namespace ee
